### PR TITLE
feat: add isValidHostname function 

### PR DESCRIPTION
**Pull request by srchase**
Adds validHostname function for checking hostname validity.  To be used for host prefixing. Ref: awslabs/smithy-typescript/pull/104

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/util-valid-hostname/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,4 @@
+export function validHostname(hostname: string): boolean {
+  const hostPattern = /^[a-zA-Z0-9]{1}$|^[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9]$/;
```
Comment (mtdowling) [inline packages/util-valid-hostname/src/index.ts]: Shouldn't you compile this once and then reuse the regex?

Comment (srchase) [reply packages/util-valid-hostname/src/index.ts]: Using the literal (vs the RegExp constructor) should cache the regex:
```
const { validHostname } = require("/aws-sdk-js-v3/packages/util-valid-hostname");

console.time('1');validHostname("test1");console.timeEnd('1');
console.time('2');validHostname("test2");console.timeEnd('2');
```
Node returns:
```
1: 0.262ms
2: 0.011ms
```

Comment (srchase) [reply packages/util-valid-hostname/src/index.ts]: Remove standalone package and added function to protocol-http.

File: packages/protocol-http/src/validHostname.ts [hunk 2]
```diff
@@ -0,0 +1,4 @@
+export function validHostname(hostname: string): boolean {
```
Comment (trivikr) [inline packages/protocol-http/src/validHostname.ts]: use `isValidHostname`
```suggestion
export function isValidHostname(hostname: string): boolean {
```

StackOverflow discussion https://softwareengineering.stackexchange.com/questions/348649

Comment (srchase) [reply packages/protocol-http/src/validHostname.ts]: Fixed.

**Pull request merged**